### PR TITLE
Avoid Hashtable-related allocations in DataBindEngine

### PR DESCRIPTION
**Pull request by stephentoub**
## Description

The _valueConverterTable is a Hashtable that uses a struct-based key.  It's boxing every time a lookup is performed or an add is performed.

## Customer Impact

Unnecessary allocation

## Regression

No

## Testing

CI

## Risk

Minimal

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/wpf/pull/6501)

### CODE REVIEW

File: src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/Data/DataBindEngine.cs [hunk 1]
```diff
@@ -368,13 +368,14 @@ internal ViewRecord GetViewRecord(object collection, CollectionViewSource key, T
                                                         Type targetType,
                                                         bool targetToSource)
         {
-            IValueConverter result = _valueConverterTable[sourceType, targetType, targetToSource];
-
-            if (result == null)
+            ValueConverterTableKey key = new ValueConverterTableKey(sourceType, targetType, targetToSource);
+            if (_valueConverterTable.TryGetValue(key, out IValueConverter result))
```
Comment (halgab) [inline src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/Data/DataBindEngine.cs:372]: Shouldn't this be
```suggestion
            if (!_valueConverterTable.TryGetValue(key, out IValueConverter result))
```
?
```suggestion
            if (_valueConverterTable.TryGetValue(key, out IValueConverter result))
```

Comment (halgab) [inline src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/Data/DataBindEngine.cs:372]: Shouldn't this be
```suggestion
            if (!_valueConverterTable.TryGetValue(key, out IValueConverter result))
```
?
```suggestion
            if (_valueConverterTable.TryGetValue(key, out IValueConverter result))
```

Comment (stephentoub) [reply src/Microsoft.DotNet.Wpf/src/PresentationFramework/MS/Internal/Data/DataBindEngine.cs]: Yup, typo, thanks.

**Review by dipeshmsft — APPROVED** [review]

**Comment by dipeshmsft** [comment]
We have taken this PR for the current Community Test Pass. Thanks for the contribution.

**Pull request merged**